### PR TITLE
Deterministic global sampling of corpus ids

### PR DESCRIPTION
**Pull request by Waino**
Tasks (corpus ids) are now sampled in a deterministic way such that all TaskQueueManagers can be aware of which task all other devices are training.

One benefit is that it is no longer necessary to communicate ready_t during gradient sync. This simplifies the algorithm, but does not seem to have any performance impact.

The PR includes refactoring of multiple locations in the code to use a joint representation of DistributedComponents, to determine which parameters are on which devices, and how they need to be communicated.
- broadcast of initial model parameters
- gradient communication during training
- optimizer

There are additional places that could be refactored in future work:
- model_builder
- module splitter

This PR also includes a SimpleLookAheadBucketing, which provides dynamic minibatching with guarantees on the maximum batch size. This allows increasing the minibatch size without causing VRAM OOM.

### CODE REVIEW

File: mammoth/bin/train.py
Comment (TimotheeMickus) [inline mammoth/bin/train.py:228]: maybe add a comment to stress this is a non-distributed context conditional branch

File: mammoth/distributed/communication.py
Comment (TimotheeMickus) [inline mammoth/distributed/communication.py:39]: "managed" ?  I'm not sure about the clarity of the terminology here, could you elaborate on that?

File: mammoth/distributed/communication.py
Comment (TimotheeMickus) [inline mammoth/distributed/communication.py:59]: good to know the hook can go, but bad for docstring. Move to a comment block outside the function / document this in the PR

File: mammoth/opts.py
Comment (TimotheeMickus) [inline mammoth/opts.py:1025]: bucket is a terrible name for that?

File: mammoth/inputters/dataloader.py
Comment (TimotheeMickus) [inline mammoth/inputters/dataloader.py:283]: please only denumericalize if logging level is set to debug?

File: mammoth/opts.py
Comment (TimotheeMickus) [inline mammoth/opts.py:1246]: so no support for max len padding, which is the thing that seems to work the least bad for now?

File: mammoth/utils/optimizers.py
Comment (TimotheeMickus) [inline mammoth/utils/optimizers.py:55]: feels like we're missing a class constructor signature inspection + kwargs popping approach

**Review by TimotheeMickus — APPROVED** [review]
LGTM, but I'm not sure about dropping the sentence-level batching + max len padding trick, which thus far is the one with the best tok/s, right?
to what extent could this be added back in, given the current implementation?

**Review by TimotheeMickus — CHANGES_REQUESTED** [review]
(Apparently issues with logging the TQM on this branch. Has this been smoketested / tested? if so I'll dismiss this second review)


```
[2024-05-06 11:46:02,478 1083  INFO] world_size = 4, queue_size = 40
[2024-05-06 11:46:02,481 1083  INFO] in task_queue_manager: node_rank 0 local_rank 0
Traceback (most recent call last):
  File "/home/nloppi/Tiedemann_project/mammoth/mammoth/train.py", line 6, in <module>
    main()
  File "/home/nloppi/Tiedemann_project/mammoth/mammoth/mammoth/bin/train.py", line 304, in main
    train(opts)
  File "/home/nloppi/Tiedemann_project/mammoth/mammoth/mammoth/bin/train.py", line 245, in train
    logger.info(f'TaskQueueManager: {global_task_queue_manager}')
  File "/home/nloppi/Tiedemann_project/mammoth/mammoth/mammoth/distributed/tasks.py", line 310, in __repr__
    kwargs = ',\n '.join(
  File "/home/nloppi/Tiedemann_project/mammoth/mammoth/mammoth/distributed/tasks.py", line 311, in <genexpr>
    f'{key}={pformat(self.__getattribute__(key))}'
AttributeError: 'TaskQueueManager' object has no attribute 'node_rank'
``

**Comment by Waino** [comment]
> LGTM, but I'm not sure about dropping the sentence-level batching + max len padding trick, which thus far is the one with the best tok/s, right? to what extent could this be added back in, given the current implementation?

The sentence-level batching + max len padding trick is implemented as a special case of the spiral bucketing data loader.

If we want both dynamic batching and the sentence-level batching + max len padding trick, then the easiest way is to keep all 3 implementations. However, it would be quite confusing to have two different ways of doing dynamic batching, of which one is known to be flaky. So maybe the dynamic spiral bucketing would be there but inaccessible through the config.

A better solution would be to refactor the sentence-level batching + max len padding trick so that it doesn't need the spiral bucketing data loader.

For benchmarking, it was just easier to rip it out.

File: mammoth/distributed/communication.py
Comment (Waino) [reply mammoth/distributed/communication.py:39]: We should rename this, if we can come up with a more descriptive name.

Maybe `externally_managed_reduce_and_rescale_grads` would do the trick?
The idea is that `only_ready` uses the metadata provided by the `has_grad` hook to make its own decisions about how to communicate certain parameters, but the new implementation needs to be managed by something else, with the caller passing in the `has_local_gradient` and `gradient_norm`.

File: mammoth/opts.py
Comment (Waino) [reply mammoth/opts.py:1025]: Bucket is indeed a terrible name. I reused the existing parameter without renaming it, while totally changing its meaning. Classic mammoth.

The plan was to either clean this up or throw this away based on the benchmark results, but unfortunately they were somewhat inconclusive.

File: mammoth/utils/optimizers.py
Comment (Waino) [inline mammoth/utils/optimizers.py:205]: Here we also use the term `managed` with the same meaning.
The caller must supply gradient syncs (a sequence of ` DistributedComponentGradientSync`. Should add a type annotation) which determine whether or not to step each suboptimizer.

**Comment by Waino** [comment]
> (Apparently issues with logging the TQM on this branch. Has this been smoketested / tested? if so I'll dismiss this second review)

In commit fbe4f5c8a1f14cab9eb428cc1504539e07fb3893
I tried to reduce the log spam somewhat, and one of the actions was to only log a single TQM (the global one) instead of each device logging their own. It seems that a local TQM could be represented, but the `__repr__` of the global one was broken.

I fixed the bug.

**Review by TimotheeMickus — APPROVED** [review]

**Comment by TimotheeMickus** [comment]
fix linting and this should be GTG

**Comment by TimotheeMickus** [comment]
LGTM! will let you the honors of merging :']

**Comment by Waino** [comment]
Closes #8

**Pull request merged**
